admin/src/react-components/service-editor.js

```diff
@@ -240,42 +261,66 @@ class ConfigurationEditor extends Component {
     }
 
     return (
-      <TextField
-        key={displayPath}
-        id={displayPath}
-        label={descriptor.name || displayPath}
-        inputProps={{ maxLength: 4096 }}
-        value={currentValue || ""}
-        onChange={ev => {
-          if (descriptor.type === "json") {
-            if (!isValidJSON(ev.target.value)) {
-              const warningMessage = `Invalid JSON for ${descriptor.name || displayPath}. See console for details.`;
-              if (this.state.warningMessage !== warningMessage) {
-                this.setState({ warningMessage });
-              }
-              console.error(`Invalid JSON for ${descriptor.name || displayPath}.`);
-              console.error(ev.target.value);
-            } else {
-              if (this.state.warningMessage !== null) {
-                this.setState({ warningMessage: null });
+      <div className={this.props.classes.longInput} key={displayPath}>
+        <TextField
+          multiline
+          rows={20}
+          key={displayPath}
+          id={displayPath}
+          label={name || displayPath}
+          inputProps={{ maxLength: 4096 }}
+          value={currentValue || (isTheme && !this.state.isDirty && JSON.stringify(theme, null, 2)) || ""}
+          onChange={ev => {
+            const value = ev.target.value;
+            if (isJson) {
+              if (!isValidJSON(value)) {
+                const warningMessage = `Invalid JSON for ${name || displayPath}. See console for details.`;
+                if (this.state.warningMessage !== warningMessage) {
+                  this.setState({ warningMessage });
+                }
+                console.error(`Invalid JSON for ${name || displayPath}.`);
+                console.error(value);
+              } else {
+                if (this.state.warningMessage !== null) {
+                  this.setState({ warningMessage: null });
+                }
               }
             }
-          }
-          this.onChange(path, ev.target.value);
-        }}
-        onBlur={ev => {
-          if (descriptor.type === "json" && isValidJSON(ev.target.value)) {
-            // Pretty print json strings
-            const pretty = JSON.stringify(JSON.parse(ev.target.value), null, 2);
-            this.onChange(path, pretty);
-          }
-        }}
-        helperText={descriptor.description}
-        type="text"
-        fullWidth
-        multiline
-        margin="normal"
-      />
+            this.onChange(path, value);
+          }}
+          onBlur={ev => {
+            const value = ev.target.value;
+            if (isJson && isValidJSON(value)) {
+              // Pretty print json strings
+              const pretty = JSON.stringify(JSON.parse(value), null, 2);
+              this.onChange(path, pretty);
+            }
+          }}
+          helperText={description}
+          type="text"
+          fullWidth
+          margin="normal"
+        />
+
+        {isTheme && (
```

Comment:
I'm thinking this still might be good here. I have ideas on how to make this more flexible in the future. but because the button ties in so much with the inputs hooks and current value it seems logical to keep in the component for now.

Reply:
I don't know how to lift the `currentValue` out from the `TextField` but in theory that's how I would _want_ to handle this. Conceptually this part is meant to be a presentational component. When I added the json validation way back when, I didn't know how to do that logic at the higher level so I just stuffed it in here. If you figure out how to handle the logic in a container component, please lift both the code you added (the button and themes specific stuff) AND the JSON validation stuff I added a while back.

But I don't mind if you don't do that now.